port of magic_gen.py to magic_gen.go

Pac23 opened the pull request:
diverged from large pr to this. solves #16

adrianreber commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,116 @@
+	"reflect"
+	"strconv"
+	"strings"
+	//"io/ioutil"
```
Please remove all uncommented code from the file, no need to keep that in git.

adrianreber commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,116 @@
+func main() {
+
+	if len(os.Args) != 3 {
+		//fmt.Println("Usage: magic-gen.py path/to/image.h path/to/magic.json")
```
This looks like a useful message to print out

adrianreber commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,116 @@
+	}
+	magic_c_header := os.Args[1]
+	magic_json := os.Args[2]
+	infile, err := os.Open(magic_c_header) // opens magic file
```
Please no comments at the end of the line. Move it on its own life above.

adrianreber commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,116 @@
+		}
+		if strings.Contains(k, "_MAGIC") == true {
+			k = strings.Replace(k, "_MAGIC", "", -1)
+			//fmt.Println(k)
```
Remove that or add a debug mode.

adrianreber commented:
Please add a Makefile to build your tool and make sure it is also built during a Travis run.

adrianreber commented:
Also add some information to the commit message about this tool. Why we need it and how to use it.

adrianreber commented:
You could also run a minimal test during Travis with a minimal input file and compare the output json.

adrianreber commented:
Please rebase to take your Travis fixes into this PR.

As long as the code is not yet committed to the main repository please no fix up commits. Just fix the existing commit. In this case you need to squash both commits to one commit and force push to your branch.

Pac23 commented:
> Also add some information to the commit message about this tool. Why we need it and how to use it.

Done 

> Please rebase to take your Travis fixes into this PR.
> 
> As long as the code is not yet committed to the main repository please no fix up commits. Just fix the existing commit. In this case you need to squash both commits to one commit and force push to your branch.

Fixed,squashed with parent.

adrianreber commented:
There are still two commented out lines in the file: fmt.Println() , please remove them

Please edit the commit message to not look like two commits messages. Make one commit message out of it. Stay under 80 characters per line in the commit message (except for error messages or examples). Add an example how you can use your program.

Have a look at https://github.com/checkpoint-restore/go-criu/blob/master/Makefile and add a step to build and lint your program. Add a test run with a minimal input (maybe only two lines) which you can easily compare maybe in the Makefile. Add the built, the lint and the test to .travis.yml.

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+func main() {
+
+	if len(os.Args) != 3 {
+		fmt.Println("Usage: magic-gen.py path/to/image.h path/to/magic.json")
```
It's no longer `.py` I guess.

kolyshkin replied:
.. and if you want to call it `magic-gen`, put it into `magic-gen` directory.

Pac23 replied:
created a magic-gen directory.

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+		txtlines = append(txtlines, scanner.Text())
+	}
+
+	for _, eachline := range txtlines {
```
Please combine these two `for` into one. IOW there is no need to have `txtlines`.

Pac23 replied:
fixed it.

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+	for scanner.Scan() {
+		txtlines = append(txtlines, scanner.Text())
+	}
+
```
`scanner.Error()` check is missing here.

Pac23 replied:
error check added

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+			continue
+		}
+
+		if itemExists(split, "#define") == false {
```
nit: `if !itemExists(split, "#define")`

Pac23 replied:
done

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+
+		key := split[1]
+		value := split[2]
+		//fmt.Println(reflect.TypeOf(value))
```
debug comment? should be removed

Pac23 replied:
removed all debug comments

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+		key := split[1]
+		value := split[2]
+		//fmt.Println(reflect.TypeOf(value))
+		if funk.Contains(all_magic, value) == true {
```
nit: remove ` == true`

Pac23 replied:
fixed

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	for k, v := range magic {
+		//fmt.Println(reflect.TypeOf(v))
```
debug comment? please remove

Pac23 replied:
removed comment

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+
+	for k, v := range magic {
+		//fmt.Println(reflect.TypeOf(v))
+		if v == "0x0" || v == "1" || k == "0x0" || v == "1" {
```
Double check for `v == "1" ` does not make sense

Pac23 replied:
removed and changed to single check

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+		}
+		vhexedv := hex2int(v)
+		check(err)
+		//fmt.Println(vhexedv)
```
debug print

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+	}
+	mapper["by_name"] = by_name
+	mapper["by_val"] = by_val
+	mapjsoned, err := json.Marshal(mapper)
```
this is very inefficient. Please use json.NewEncoder() and json.Encode()

Pac23 replied:
slightly inefficient but json.Encode spits out digits,this is more human readable.Although it doesn't really matter to us as have to unmarshal/decode alter anyways.

Pac23 replied:
marshal changed to json.Encode()

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+
+// function to convert hex string to int
+func hex2int(hexStr string) uint64 {
+	// remove 0x suffix if found in the input string
```
prefix?

Pac23 replied:
sorry typo

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+// function to convert hex string to int
+func hex2int(hexStr string) uint64 {
+	// remove 0x suffix if found in the input string
+	cleaned := strings.Replace(hexStr, "0x", "", -1)
```
strconv.ParseUint() is perfectly capable of parsing a string with `0x` prefix, see https://golang.org/pkg/strconv/#ParseInt

Pac23 replied:
tried that,returns 0 stating invalid no. Thus have to clean the prefix first then send to strconv.ParseUint

kolyshkin commented on crit-go/scripts/magic_gen.go:
```diff
@@ -0,0 +1,115 @@
+		return true
+	} else {
+		return false
+	}
```
`return s.Index(0).Interface() == item`

Pac23 replied:
done

adrianreber commented on crit-go/magic-gen/magic_gen.go:
```diff
@@ -0,0 +1,105 @@
+// function to convert hex string to int
+func hex2int(hexStr string) uint64 {
+	// remove 0x prefix if found in the input string
+	cleaned := strings.Replace(hexStr, "0x", "", -1)
```
is this really necessary? Can't Go convert from 0x123e to integer?

Pac23 replied:
strconv.ParseInt considers it to be a long string thus ends with "0" or invalid syntax [playground](https://play.golang.org/p/mwtttfzROIW) thus the step of cleaning it first seems to be the only option to get around the problem.

adrianreber replied:
What about `n, err := strconv.ParseUint(s, 0, 64)`? That works for me. If you set the base to `0` it works correctly.

adrianreber commented on crit-go/magic-gen/magic_gen.go:
```diff
@@ -0,0 +1,105 @@
+)
+
+func main() {
+	// checks aregyument correctness
```
typo in comment

adrianreber commented on crit-go/magic-gen/magic_gen.go:
```diff
@@ -0,0 +1,105 @@
+		by_name[k] = vhexedv
+		by_val[vhexedv] = k
+	}
+	// by_val and by_name ampped to a master map to be converted to json
```
typo in comment

adrianreber commented:
```
$ make
panic: open magic.h: no such file or directory

goroutine 1 [running]:
main.check(...)
	/share/go/src/github.com/checkpoint-restore/go-criu/crit-go/magic-gen/magic_gen.go:83
main.main()
	/share/go/src/github.com/checkpoint-restore/go-criu/crit-go/magic-gen/magic_gen.go:24 +0xba2
exit status 2
make: *** [makefile:21: test/magicgen] Error 1
```

adrianreber commented on .travis.yml:
```diff
@@ -23,3 +24,7 @@ script:
+  # This builds crit-go
+  - make -C crit-go/magic-gen lint build magicgen test
+  # Run test of crit-go
+  - sudo make test
```
This line is not needed, You are already running the test in line 28. Remove the last two lines.

adrianreber commented on crit-go/magic-gen/Makefile:
```diff
@@ -0,0 +1,53 @@
+test: test/magicgen
+		
+magicgen: build
+	curl -s https://raw.githubusercontent.com/checkpoint-restore/criu/criu-dev/criu/include/magic.h -o magic.h
```
put the download in its own target and make this target dependent on it. This way the file will not be redownloaded for every step:
```
magic.h:
        curl -s https://raw.githubusercontent.com/checkpoint-restore/criu/criu-dev/criu/include/magic.h -o magic.h

magicgen: build magic.h
	./magic-gen magic.h magic.json
	@rm -f magic.h
```

adrianreber commented on crit-go/magic-gen/Makefile:
```diff
@@ -0,0 +1,53 @@
+build:
+	 @$(GO) build -v
+
+test/magicgen:
```
Add a dependency on build:
```
test/magicgen: build
```

adrianreber commented on crit-go/magic-gen/Makefile:
```diff
@@ -0,0 +1,53 @@
+	test_output=`mktemp`
+	echo "#define TEST_MAGIC 0x12345678" > test_input
+	@rm -f test_output 
+	@$(GO) run magic-gen.go test_input test_output
```
Use the binary built by `build`.

./magic-gen test_input test_output

adrianreber commented on crit-go/magic-gen/Makefile:
```diff
@@ -0,0 +1,53 @@
+
+test/magicgen:
+	test_input=`mktemp`
+	test_output=`mktemp`
```
Remove the two mktemp lines as you are not using the variables below anyway. You are hard-coding the input and output to `test_input` and `test_output`, which is probably okay here.

adrianreber commented on crit-go/magic-gen/Makefile:
```diff
@@ -0,0 +1,53 @@
+	@rm -f testmagic.json
+	@rm -f test_output
+
+.PHONY: build test clean lint magicgen
```
add a newline at the end of the file

adrianreber commented on crit-go/magic-gen/magic-gen.go:
```diff
@@ -0,0 +1,103 @@
+func main() {
+	// checks argument correctness
+	if len(os.Args) != 3 {
+		fmt.Println("Usage: magic_gen.go path/to/magic.h path/to/magic.json")
```
I think the resulting binary is called `magic-gen` and not  `magic_gen.go`. Adapt the help output.

adrianreber commented:
Nice, with integrated travis tests. Looks good. I added a couple more reviews, but then it should be ready to be merged.

adrianreber commented:
Thanks! Merged! Great work!
